### PR TITLE
Final feature update

### PR DESCRIPTION
**Pull request by ntninja**
After several big PRs to this project, here's the last one!

This implements all the remaining features I have identified as useful for py-ipfs (filesystem stats tracking, .put create and replace parameters, .put_new & .rename) – plus a bunch of cleanups and bug fixes.

From here on in, I only expect to contribute smaller fixes, performance improvements and, hopefully lot's of, docs. Also, `.query` is still in a sorry state and I do fell like fixing it, so maybe we should just drop it (it would make a lot more sense if we had a DB backend to test it with)?

Looking forward to your review!

### CODE REVIEW

File: datastore/core/binarystore.py [hunk 1]
```diff
@@ -256,6 +256,33 @@ class Datastore(trio.abc.AsyncResource):
...
 	
+	async def rename(self, key1: key_.Key, key2: key_.Key, *, replace: bool = True) -> None:
+		"""Moves the data at name *key1* to *key2*
+		
+		Arguments
+		---------
+		key1
+			The key to rename, must exist
+		key2
+			The new name of the key, if *replace* is ``False`` this must not exist
```
Comment (AliabbasMerchant) [inline datastore/core/binarystore.py]: Can we change this to: "The new name of the key, if *replace* is ``False`` this must be None"?  Does not exist sounds like the argument can be omitted, which would be a TypeError of missing positional arguments.

File: datastore/core/binarystore.py [hunk 2]
```diff
@@ -481,6 +508,31 @@ def _collection(self, key: key_.Key, parent: bool = True) \
...
```
Comment (AliabbasMerchant) [inline datastore/core/binarystore.py]: Can we change this to: "The new name of the key, if *replace* is ``False`` this must be None"?  Does not exist sounds like the argument can be omitted, which would be a TypeError of missing positional arguments.

File: datastore/core/binarystore.py [hunk 3]
```diff
@@ -704,6 +757,36 @@ def __init__(self, datastore: Datastore):
...
```
Comment (AliabbasMerchant) [inline datastore/core/binarystore.py]: Can we change this to: "The new name of the key, if *replace* is ``False`` this must be None"?  Does not exist sounds like the argument can be omitted, which would be a TypeError of missing positional arguments.

File: datastore/core/objectstore.py [hunk 4]
```diff
@@ -286,6 +286,33 @@ class Datastore(typing.Generic[T_co], trio.abc.AsyncResource):
...
 	
+	async def rename(self, key1: key_.Key, key2: key_.Key, *, replace: bool = True) -> None:
+		"""Moves content at name *key1* to *key2*
+		
+		Arguments
+		---------
+		key1
+			The key to rename, must exist
+		key2
+			The new name of the key, if *replace* is ``False`` this must not exist
```
Comment (AliabbasMerchant) [inline datastore/core/objectstore.py]: Can we change this to: "The new name of the key, if *replace* is ``False`` this must be None"?  Does not exist sounds like the argument can be omitted, which would be a TypeError of missing positional arguments.

File: datastore/core/objectstore.py [hunk 5]
```diff
@@ -771,6 +824,37 @@ def query(self, query):
...
 	
+	async def rename(self, key1: key_.Key, key2: key_.Key, *, replace: bool = True) -> None:
+		"""Moves the content at name *key1* to *key2*
+		
+		Arguments
+		---------
+		key1
+			The key to rename, must exist
+		key2
+			The new name of the key, if *replace* is ``False`` this must not exist
```
Comment (AliabbasMerchant) [inline datastore/core/objectstore.py]: Again, same suggestion

File: datastore/core/serialize.py [hunk 6]
```diff
@@ -213,6 +213,32 @@ def __init__(self, datastore: binarystore.Datastore,
...
 	
+	async def rename(self, key1: key_.Key, key2: key_.Key, *, replace: bool = True) -> None:
+		"""Moves the content at name *key1* to *key2*
+		
+		Arguments
+		---------
+		key1
+			The key to rename, must exist
+		key2
+			The new name of the key, if *replace* is ``False`` this must not exist
```
Comment (AliabbasMerchant) [inline datastore/core/serialize.py]: Again, same suggestion

File: datastore/filesystem/filesystem.py [hunk 7]
```diff
@@ -1125,6 +1126,75 @@ def _delete_sync(self, path: typing.Union[os_PathLike_str, str]) -> None:
...
+	async def rename(self, key1: datastore.Key, key2: datastore.Key, *,
+	                 replace: bool = True) -> None:
+		"""Moves key *key1* to *key2*
+		
+		Arguments
+		---------
+		key1
+			The key to rename, must exist
+		key2
+			The new name of the key, if *replace* is ``False`` this must not exist
```
Comment (AliabbasMerchant) [inline datastore/filesystem/filesystem.py]: Again, same suggestion

**Review by AliabbasMerchant — COMMENTED** [review]
Sorry @alexander255 , your code is waaaaay above my level 😢 😅 

I tried to review it, but can't be so sure. Can someone else also review it please?

File: datastore/core/binarystore.py (hunk 1 above)
Comment (ntninja) [reply datastore/core/binarystore.py]: Actually it's supposed to mean that the target key may not already exist if the *replace* parameter is False (cause if it did exist it would have to be replaced). As a matter of fact, this parameter *may* not be `None`, that wouldn't make any sense for a rename :upside_down_face:.

Do have any idea for a concise description for capturing the above meaning?

**Comment by ntninja** [comment]
@AliabbasMerchant: I literally spent over a week (not in one go of course) in writing and testing that code until I felt confident enough to share – concurrency is just so darn hard, but I guess I just like the challenge :sweat_smile: 

*Some thoughts on review size*:

Review in this case is really intended to be done on a commit by commit basis, but even then the diffs are unfortunately quite large and require some effort to understand and analyze. Considering this and our general lack of manpower I realize a full review isn't actually realistic, so I'll leave this open for comments for a week or two and then merge it (I feel like that has been our effective practice with this anyways).

TL;DR: So please anybody who feels like taking a stab at this, please comment on anything you find, but don't feel obliged to read *everything* to leave feedback. Anything is valuable!

And as I said, I don't expect to do such large diffs in the future.

File: datastore/core/binarystore.py (hunk 1 above)
Comment (AliabbasMerchant) [reply datastore/core/binarystore.py]: Oh. Right.  
Can we phrase it as: "The new name of the key. If *replace* is ``False``, a key of the same name should not already exist"?

**Comment by AliabbasMerchant** [comment]
@alexander255 ,
sorry for the late reply.

For me, its the typing part that is more difficult than the concurrency part 😅 
I understand that you have put in lots of efforts to complete this PR, and I would like to replicate that. Give me a few more days, and I will completely review it, by hook or by crook.

**Comment by ntninja** [comment]
> For me, its the typing part that is more difficult than the concurrency part 😅

Yeah, that turned out being quite messy overall. 😦
Most of it could be avoided by duplicating all binary vs object code (like done for the Object/BinaryDatastore base types) everywhere, but that's very bad practice and leads to unnecessary bit-rot. So I made the best of the few type-level abstractions mypy offers, but they currently just aren't good enough and so things wind up somewhat ugly and repetitive. – It feels like it would be such a waste to give up on it now, that I got it to “mostly work most places” however. I do believe it can be replaced with better code in the future though. 🙂 

> I understand that you have put in lots of efforts to complete this PR, and I would like to replicate that. Give me a few more days, and I will completely review it, by hook or by crook.

Thank you! I do appreciate your feedback on this stuff!

File: datastore/core/binarystore.py (hunk 1 above)
Comment (ntninja) [reply datastore/core/binarystore.py]: Thank you! Applied wording similar to yours.

File: datastore/adapter/mount.py [hunk 8]
```diff
@@ -12,28 +12,25 @@
 ]
 
 
+mount_item_t = typing.Tuple['mount_tree_t[DS]', typing.Optional[DS]]
+
+
+class mount_tree_t(typing.Dict[str, mount_item_t[DS]]):
+	...
+
```
Comment (AliabbasMerchant) [inline datastore/adapter/mount.py:20]: This was a nice decision!

File: tests/adapter/test_tiered.py [hunk 9]
```diff
@@ -86,3 +87,60 @@ async def test_tiered(DatastoreTests, Adapter, DictDatastore, encode_fn):
...
+		assert not await s1.contains(k2)
+		assert not await s2.contains(k1)
+		assert await s2.contains(k2)
+		
+		# Ensure red back value didn't change
+		assert await ts.get_all(k1) == encode_fn("1")
+		assert await ts.get_all(k2) == encode_fn("1")
+		
+		assert await s1.contains(k1)
+		assert await s1.contains(k2)
```
Comment (AliabbasMerchant) [inline tests/adapter/test_tiered.py:131]: I may be wrong (coz the tests are passing), but shouldn't this be `assert await not s1.contains(k2)`?

**Comment by AliabbasMerchant** [comment]
@alexander255 I went through the PR again. Couldn't find anything amiss. Lets merge it.

**Comment by AliabbasMerchant** [comment]
No one:  

@alexander255 🙌 :  
- [x] Writes async-friendly code
- [x] Writes type-hints even in dynamically typed languages
- [x] Patiently waits for PR review

File: tests/adapter/test_tiered.py (hunk 9 above)
Comment (ntninja) [reply tests/adapter/test_tiered.py:131]: Just read through it again and can confirm that this behaviour is intended: The idea behind the tiered datastore is to accelerate access to commonly used data – it does this by storing data read from a lower store (in practice this could be a slower disk or a remote database) in all the datastores above (like fast in-memory or on-disk caches) to accelerate access to it on subsequent reads. Hence the behaviour tested here is correct for this particular datastore (it would have been different had we used `.get_all` on the underlying datastores directly).

**Comment by ntninja** [comment]
Thanks! I'm not always as patient as you know, but as other things have been demanding my attention recently and there was no hurry with this one – why not wait for the commentary? :slightly_smiling_face: 
Thank you for taking the time to have a long hard look at this, I know its been a lot! :raised_hands:
